source/extensions/filters/http/jwt_authn/extractor.cc

```
@@ -0,0 +1,154 @@
+};
+
+// The JwtLocation for header extraction.
+class JwtHeaderLocation : public JwtLocationBase {
+public:
+  JwtHeaderLocation(const std::string& token, const std::unordered_set<std::string>& issuers,
+                    const LowerCaseString& header)
+      : JwtLocationBase(token, issuers), header_(header) {}
+
+  void removeJwt(Http::HeaderMap& headers) const override { headers.remove(header_); }
+
+private:
+  // the header name the JWT is extracted from.
+  const LowerCaseString& header_;
+};
+
+// The JwtLocation for param extraction.
+class JwtParamLocation : public JwtLocationBase {
+public:
+  JwtParamLocation(const std::string& token, const std::unordered_set<std::string>& issuers,
+                   const std::string&)
+      : JwtLocationBase(token, issuers) {}
+
+  void removeJwt(Http::HeaderMap&) const override {
+    // TODO(qiwzhang): remove JWT from parameter.
+  }
+
+private:
+};
+
+} // namespace
+
+Extractor::Extractor(const JwtAuthentication& config) {
+  for (const auto& rule : config.rules()) {
+    for (const auto& header : rule.from_headers()) {
+      addHeaderConfig(rule.issuer(), LowerCaseString(header.name()), header.value_prefix());
+    }
+    for (const std::string& param : rule.from_params()) {
+      addParamConfig(rule.issuer(), param);
+    }
+
+    // If not specified, use default locations.
+    if (rule.from_headers().empty() && rule.from_params().empty()) {
+      addHeaderConfig(rule.issuer(), Http::Headers::get().Authorization,
+                      JwtConstValues::get().BearerPrefix);
+      addParamConfig(rule.issuer(), JwtConstValues::get().AccessTokenParam);
+    }
+  }
+}
+
+void Extractor::addHeaderConfig(const std::string& issuer, const LowerCaseString& header_name,
+                                const std::string& value_prefix) {
+  std::string map_key = header_name.get() + value_prefix;
+  auto& map_value = header_maps_[map_key];
+  if (!map_value) {
```

Should we raise an exception and reject the config if there are dupes header locations, or do you prefer to just silently ignore later ones because it doesn't matter (or does it, for ordering reasons?!)?

---

We should support different issuers with same header/prefix.  

For example,  two issuers, issuer1 and issuer2 can extract their token from "Authorization: Bearer".   

A request with headers:
Authroization: Bearer <issuer1_token>

it is OK, it will use issuer1 config to process.

Similarly,  it should work for:
Authroization: Bearer <issuer2_token>